### ISSUE TITLE
[Help] Error loading a dynamic library

### ISSUE THREAD

**Issue by Kiwi-sia** [exelix11/SysDVR#208]
**Where do you get stuck ?**
I completed the Linux part of the guide and got everything installed:
![installed](https://user-images.githubusercontent.com/81862614/202928666-f15c7730-ad8d-48b4-9d94-b51d307ab675.png)
However, I can't run the client on my PC at all (see the next section)

**What's the error message you get ?**
![sysdvr-client error](https://user-images.githubusercontent.com/81862614/202928640-b4b2e6e5-09b9-4520-849b-c80f1772ea41.png)
The message is: 
```
There was an error loading a dynamic library. Make sure you installed all the dependencies and that you have the correct version of the libraries.
If all libraries are properly installed ensure their path is set in your dynamic loader environment variable (PATH on Windows, LD_LIBRARY_PATH on Linux and DYLD_LIBRARY_PATH on MacOS)

In case of problems specific to ffmpeg or libavcodec you can override the loader path by adding to the command line --libdir <library path> where <library path> is the folder containing the dyamic libraries of ffmpeg (or symlinks to them).
For example on x64 linux you can try --libdir /lib/x86_64-linux-gnu/

Full error message:
System.DllNotFoundException: Unable to load DLL 'avcodec.59 under ': The specified module could not be found.
   at FFmpeg.AutoGen.ffmpeg.LoadLibrary(String libraryName, Boolean throwException)
   at FFmpeg.AutoGen.ffmpeg.<>c.<.cctor>b__7_0(String libraryName)
   at FFmpeg.AutoGen.ffmpeg.<>c.<.cctor>b__7_18()
   at FFmpeg.AutoGen.ffmpeg.av_packet_alloc()
   at SysDVR.Client.Player.H264StreamTarget..ctor()
   at SysDVR.Client.Player.PlayerManager..ctor(Boolean HasVideo, Boolean HasAudio, Boolean hwAcc, String codecName, String quality)
   at SysDVR.Client.Program.HandleStreamingCommands()
   at SysDVR.Client.Program.ProgramMain()
   at SysDVR.Client.Program.Main(String[] args)
```

**Setup information**
PC OS: Ubuntu 20.04
I am not sure if my console is even relevant since I can't get it started on the PC end but still...
 - Console firmware: 14.1.2
 - Custom firmware name and version: Atmosphere (wasn't able to find the version... Will update if needed)
 - SysDVR version: 5.3.1

**Additional context**
I tried adding '--libdir /lib/x86_64-linux-gnu/' to the command but it gives the exact same error message.
I am pretty sure I just installed everything 'as-is' and didn't change any directories or dependencies

**Comment by exelix11** [comment]
you should install ffmpeg from your package manager, making sure it is version 59.
If it's not you will have to google how to install libavcodec 59 on your distro.

**Comment by Kiwi-sia** [comment]
It shows that **libavcodec 59** is installed, not so sure about **ffmpeg** version though (it seems to be 5.1.2-0ubuntu1~20.04.sav1, whatever that means)
![libvacodec](https://user-images.githubusercontent.com/81862614/202929341-fb1d00da-cf5a-489f-bae9-a277a813e81d.png)

**Comment by exelix11** [comment]
It should be fine, i just say ffmpeg cause it's easier to explain and will pull all the other dependencies automatically.
You should run a find to search for the avcodec.59 so file and use its path as --libdir
in case you have a file called avcodec.59.so.1 or other numbers you should symlink it to just avcodec.59.so iirc

**Comment by Kiwi-sia** [comment]
It is in the /lib/x86_64-linux-gnu (as the error hint suggested), there is also a link to the .59 file.
![linked](https://user-images.githubusercontent.com/81862614/202930616-2dc6c00b-69c7-4230-96dc-5590f0d554e1.png)

I still cannot make SysDVR client run :C
![same error](https://user-images.githubusercontent.com/81862614/202930683-27c34f1c-c362-47e6-9f72-39d012d9be8b.png)

**Comment by exelix11** [comment]
try making a link called libavcodec.59.so, if that doesn't work you might have to ptrace it figure out what path is it looking for and failing to find (could be a dependency)

**Comment by Kiwi-sia** [comment]
Link 'libavcodec.59.so' didn't work. I have no idea how to ptrace...

**Comment by exelix11** [comment]
I went to check ffmpeg bindings source code and i think there's a [bug](https://github.com/Ruslan-B/FFmpeg.AutoGen/blob/534a3eb0ff295bc9bc8382b311df407180ab5051/FFmpeg.AutoGen/FunctionResolverBase.cs#L74) that affects linux and mac.

Can you try the following:
- Create a symlink called just `avcodec.so.59`
- Launch sysdvr with `--libdir /../../../../../lib/x86_64-linux-gnu`

I suspect the library is concatenating the `lib` part at the start of the path breaking everything, .. will trick it into loading from the proper path, if this fixes it i'll open an issue on their repo.

Wither way i think the proper solution in your case is to update the LD_LIBRARY_PATH folder to include /lib/x86_64-linux-gnu

**Comment by Kiwi-sia** [comment]
```
kiwi@kiwi:~/Downloads/SysDVR-Client$ dotnet SysDVR-Client.dll --libdir /../../../../../lib/x86_64-linux-gnu
SysDVR-Client - SysDVR 5.3.1 by exelix
https://github.com/exelix11/SysDVR 

There was an error loading a dynamic library. Make sure you installed all the dependencies and that you have the correct version of the libraries.
If all libraries are properly installed ensure their path is set in your dynamic loader environment variable (PATH on Windows, LD_LIBRARY_PATH on Linux and DYLD_LIBRARY_PATH on MacOS)

In case of problems specific to ffmpeg or libavcodec you can override the loader path by adding to the command line --libdir <library path> where <library path> is the folder containing the dyamic libraries of ffmpeg (or symlinks to them).
For example on x64 linux you can try --libdir /lib/x86_64-linux-gnu/

Full error message:
System.DllNotFoundException: Unable to load DLL 'avcodec.59 under /../../../../../lib/x86_64-linux-gnu': The specified module could not be found.
   at FFmpeg.AutoGen.ffmpeg.LoadLibrary(String libraryName, Boolean throwException)
   at FFmpeg.AutoGen.ffmpeg.<>c.<.cctor>b__7_0(String libraryName)
   at FFmpeg.AutoGen.ffmpeg.<>c.<.cctor>b__7_18()
   at FFmpeg.AutoGen.ffmpeg.av_packet_alloc()
   at SysDVR.Client.Player.H264StreamTarget..ctor()
   at SysDVR.Client.Player.PlayerManager..ctor(Boolean HasVideo, Boolean HasAudio, Boolean hwAcc, String codecName, String quality)
   at SysDVR.Client.Program.HandleStreamingCommands()
   at SysDVR.Client.Program.ProgramMain()
   at SysDVR.Client.Program.Main(String[] args)
Segmentation fault (core dumped)
```
The links 
![collection_avc](https://user-images.githubusercontent.com/81862614/202932126-fd068979-d850-4e56-85b5-36c05c2124a7.png)

**Comment by exelix11** [comment]
Ok so i chcked again and it seems latest sysdvr is shipping an older version of ffmpeg.autogen which doesn't have this issue. Should probably still report this to the dev but you're facing a different issue.

At this point you should debug this issue on your machine, lookup how to use ltrace, you're looking for a dlopen call in the log so you can see what path is it looking for.

**Comment by elsword50716** [comment]
Try running `brew reinstall rav1e`. it works for me.

**Comment by c01dc0ffee** [comment]
```
short@base24:~$ whereis libavcodec.so.59
libavcodec.so.59: /usr/lib/x86_64-linux-gnu/libavcodec.so.59 /usr/lib/libavcodec.so.59

short@base24:~$ ldd /usr/lib/x86_64-linux-gnu/libavcodec.so.59
	linux-vdso.so.1 (0x00007ffd553cc000)
	libswresample.so.4 => not found
	libavutil.so.57 => not found
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f5213519000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f52134ee000)
	libva.so.2 => /lib/x86_64-linux-gnu/libva.so.2 (0x00007f52134bd000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f52134a1000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5213200000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f52146d8000)
```

libavcodec 59 depends on libavutil 57 and libswresample 4. The highest version of ffmpeg for Ubuntu 22.04 is 4.4.2 and lacks these. Installing ffmpeg >=5.0 should fix this (I opted to upgrade to 22.10)
